api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto

```diff
@@ -67,8 +67,24 @@ message TcpProxy {
     option (udpa.annotations.versioning).previous_message_type =
         "envoy.config.filter.network.tcp_proxy.v2.TcpProxy.TunnelingConfig";
 
-    // The hostname to send in the synthesized CONNECT headers to the upstream proxy.
-    string hostname = 1 [(validate.rules).string = {min_len: 1}];
+    message AutoSni {
+      // Optional port used to create the request-target for all CONNECT request, i.e. CONNECT <hostname_from_sni>:<default_port>
+      // If default_port is not specified, 443 is used.
+      uint32 port = 1;
+    }
+
+    oneof hostname_config {
+      option (validate.required) = true;
+
+      // Static hostname used in request-target of CONNECT request
+      string hostname = 1;
```

Comment:
Unfortunately we can't upgrade an existing field to `oneof` per https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md

Reply:
Ok, I understand. So the only way is to deprecate existing field `hostname` and add another one, let's say `static_hostname`, and wrap it together with `auto_sni` in `oneof`? Take a look at this [comment](https://github.com/envoyproxy/envoy/issues/19612#issuecomment-1028147274).
Then the schema would be:
```
message TunnelingConfig {
  string hostname = 1 [deprecated = true];

  oneof hostname_config {
    string static_hostname = 4;
    AutoSni auto_sni = 5;
  }
}
```

Reply:
We could also go with the relaxed validation and mutual exclusion in the second part of https://github.com/envoyproxy/envoy/issues/19612#issuecomment-1028147274. Either of these is valid.

Reply:
sni is indeed a reasonable source of host name header. 

Can we relax the hostname value to %EXPR%? 
I have the use case that populating the value from %FILTER_STATE(KEY:F):Z% and %DYNAMIC_METADATA(NAMESPACE:KEY*):Z%
It could be done in another PR but I'd like to point out SNI is not only source of required value.

Reply:
> Could you update the PR to move "hostname =1" out of one_of
Done